app/controllers/x_site_auth_controller.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require 'jwt'
+
+# Allows log-in to other sites (i.e. the forums) with your DAD account.
+class XSiteAuthController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: %i[sign]
+  before_action :set_no_cache
+
+  # Does the user have an x-auth cookie set?
+  # Other sites snooping on whether a user is logged in is prevented by CORS.
+  def auto_login_available
+    render plain: (!x_auth_cookie.nil?).to_s
+  end
+
+  # Log in to this site, and get redirected back to the DAD zone with your new auth code.
+  def login
+    # `return_to` is a parameter rather than using the referer header
+    # so that disabling referers (which I will likely do globally for the DAD forums)
+    # won't break cross-site login.
+    unless params[:return_to]
+      flash[:danger] = "You didn't specify which page to return to!"
+      # Alternatively, it would be equally valid to kick back users to the home page.
+      # I don't suppose it matters that much, because these cases are unlikely to ever be hit,
+      # and render_unauthorized makes debugging easier than a redirect does.
+      render_unauthorized
+      return
+    end
+
+    @return_to = URI(params[:return_to])
+
+    # These checks aren't strictly necessary because removing them wouldn't cause a security bug
+    # due to stuff like CORS settings, but they're still avoided to prevent client-side code bugs
+    # (i.e. so that *my* code won't set the wrong return URL).
+    unless @return_to.scheme == 'https'
+      flash[:danger] =
+        "HTTPS is required for cross-site authentication; #{@return_to} is not an HTTPS URL."
+      render_unauthorized
+      return
+    end
+
+    unless @return_to.host == ENV['X_AUTH_HOST']
```

Comment:
What is X_AUTH_HOST? The website that DAD permits you to cross-authenticate login to (aka, only the DAD forums)?

Reply:
Yes.